hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OzoneManager.java

```diff
@@ -2665,6 +2665,12 @@ public List<OmVolumeArgs> listVolumeByUser(String userName, String prefix,
     try {
       metrics.incNumVolumeLists();
       if (isAclEnabled) {
+        String remoteUserName = remoteUserUgi.getShortUserName();
+        // if not admin nor list my own volumes, check ACL.
+        if (!remoteUserName.equals(userName) && !isAdmin(remoteUserUgi)) {
```

Comment:
This check is a bit weak the cli param is not guaranteed to use short user name only?

Reply:
Thanks @duongkame @kerneltime for bringing this up.

Looks like we need to convert `userName` to UGI then use its short name as well for comparison.

Reply:
The original problem is: volumes are returned to unauthorized users. Why don't we just filter returned volumes and only return those that user (remoteUserUgi) has access to?

Reply:
Something like this:

```diff
diff --git a/hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OzoneManager.java b/hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OzoneManager.java
index 67803786df..3a9308ef3f 100644
--- a/hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OzoneManager.java
+++ b/hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OzoneManager.java
@@ -2592,8 +2592,13 @@ public List<OmVolumeArgs> listVolumeByUser(String userName, String prefix,
       metrics.incNumVolumeLists();
       if (isAclEnabled) {
         String remoteUserName = remoteUserUgi.getShortUserName();
+        // Convert userName to short username
+        final UserGroupInformation ugiUserParam =
+            UserGroupInformation.createRemoteUser(userName);
+        final String userParamShortName = ugiUserParam.getShortUserName();
         // if not admin nor list my own volumes, check ACL.
-        if (!remoteUserName.equals(userName) && !isAdmin(remoteUserUgi)) {
+        if (!remoteUserName.equals(userParamShortName)
+            && !isAdmin(remoteUserUgi)) {
           omMetadataReader.checkAcls(ResourceType.VOLUME,
               StoreType.OZONE, ACLType.LIST,
               OzoneConsts.OZONE_ROOT, null, null);
diff --git a/hadoop-ozone/ozonefs-common/src/main/java/org/apache/hadoop/fs/ozone/BasicRootedOzoneClientAdapterImpl.java b/hadoop-ozone/ozonefs-common/src/main/java/org/apache/hadoop/fs/ozone/BasicRootedOzoneClientAdapterImpl.java
index c1d882b2e4..20e9f58f81 100644
--- a/hadoop-ozone/ozonefs-common/src/main/java/org/apache/hadoop/fs/ozone/BasicRootedOzoneClientAdapterImpl.java
+++ b/hadoop-ozone/ozonefs-common/src/main/java/org/apache/hadoop/fs/ozone/BasicRootedOzoneClientAdapterImpl.java
@@ -538,7 +538,17 @@ public boolean deleteObject(String path, boolean recursive)
     }
     try {
       OzoneBucket bucket = getBucket(ofsPath, false);
-      bucket.deleteDirectory(keyName, recursive);
+      if (bucket.isLink()) {
+        final String bucketName = bucket.getName();
+        final String volumeName = bucket.getVolumeName();
+        LOG.debug("removing linked bucket '{}' from volume '{}'",
+            bucketName, volumeName);
+        // Delete the bucket from volume
+        OzoneVolume vol = objectStore.getVolume(volumeName);
+        vol.deleteBucket(bucketName);
+      } else {
+        bucket.deleteDirectory(keyName, recursive);
+      }
       return true;
     } catch (OMException ome) {
       LOG.error("delete key failed {}", ome.getMessage());
```

Reply:
> The original problem is: volumes are returned to unauthorized users. Why don't we just filter volumes to only return those that the requesting user (remoteUserUgi) has access to?

Hi @duongkame , I understand your thought of returning the **intersection** of volumes that both users has LIST permission to. But IMO that would lead to further confusing behavior, and would expose some of the volumes the other user has LIST permission to (partially leaking ACL). Rejecting outright for non-admin and non-owner here is what I would prefer.

Also a note: The reason we check ACL against `OZONE_ROOT` is because of `ozone.om.volume.listall.allowed`, where we want to mimic HDFS list `/` behavior in ofs.

Reply:
Got it, thanks @smengcl .